### ISSUE TITLE
How to maintain Fullscreen mode during playlist manifest transitions ?

### ISSUE THREAD

**Issue by loicraux** [shaka-project/shaka-player#9922]
### FAQ

- [x] I have read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues

### If the question is related to FairPlay, have you read the tutorial?

Not applicable

### What version of Shaka Player are you using?

4.15.25

### What browser and OS are you using?

Chrome/Chromium, and any OS (Windows, MacOS, Linux)

### Context

I am developing a video player (Electron-based) using Shaka Player and implementing a playlist feature.

This playlist feature is implemented by programmatically unloading current item and loading the next item using unload/load APIs of Shaka.Player.

This playlist feature is for instance used to organize screenings of short films. For this purpose the player is switched to fullscreen mode (by the user itself, say on the first item of the playlist, before he hits the play button...)

### The problem is

when the player transitions from the current video to the next item in the playlist, the browser (Chromium) automatically exits fullscreen mode. This is very annoying for the screening of course. So **I'd like the player to remain in fullscreen mode when loading the next item in the playlist**, no matter if this is automatically (when the 'ended' event is sent) or triggered by a user interaction (click on button SkipToNext / SkipToPrevious).

### Question(s)

- What could explain this behavior of exiting Fullscreen mode when a new asset is loaded ?

  - Is this behavior forced by the code of Shaka Player ?
  - Or is this a built-in browser behavior (part of the Fullscreen API spec?) ?

- How could I work around this ?

  - Should I requestFullscreen on a `<div>` wrapper of the `<video>` element instead of the `<video>` element itself ? Because the `<div>` would persist and never natively "unload", Chromium would happily keep the wrapper `<div>` in fullscreen while Shaka Player tears down the inner `<video>` element's source and loads the next playlist item, maybe ?
  - Should I consider using the Preload API ? Would this be of any help to stay in fullscreen mode ? (see [issue "Built-in Playlist feature in Shaka Player"](https://github.com/shaka-project/shaka-player/issues/8508)). I guess no bu please confirm...
  - Should I make use of the SwitchingToNewStream event ? See [PR "feat: Add switchingtonewstream event"](https://github.com/shaka-project/shaka-player/pull/8532)
  - Would the "new" Queue Manager (see [PR "Add queue manager"](https://github.com/shaka-project/shaka-player/pull/8649)) be helpful ?

Thank you very much for your help !

**Comment by loicraux** [comment]
I have asked Claude the same questions (for fun). Answer in my next comment. I'm not really sure how truthful it is?

**Comment by loicraux** [comment]
### Why It Happens

The root cause is the **Fullscreen API specification itself**, not Shaka Player's code specifically. The spec states that the browser must exit fullscreen when the *fullscreen element* is "removed from the document" or undergoes certain DOM-level changes. When you call `shaka.Player.unload()`, Shaka tears down the `MediaSource` object attached to the `<video>` element (via `URL.revokeObjectURL` or clearing `srcObject`/`src`). In Chromium, **clearing or changing the `src`/`srcObject` of a `<video>` element that is itself the fullscreen element** triggers the fullscreen exit algorithm. This is Chromium's interpretation of the spec — the media pipeline is being torn down, so the browser considers the fullscreen context invalidated.

So to be precise: Shaka doesn't explicitly call `document.exitFullscreen()`. It simply does its normal cleanup during `unload()`, and Chromium reacts by exiting fullscreen because the fullscreen target (the `<video>`) lost its media source.

### The Fix: Fullscreen on a Wrapper `<div>`

Yes — **this is the correct and canonical solution**. Put fullscreen on a container `<div>`, not on the `<video>` element itself.

```html
<div id="video-container" style="width:100%; height:100%; background:#000;">
  <video id="video" style="width:100%; height:100%;"></video>
</div>
```

```js
// Enter fullscreen on the CONTAINER, not the <video>
document.getElementById('video-container').requestFullscreen();
```

The reason this works is exactly your intuition: the `<div>` is a stable DOM element that never gets torn down or modified during Shaka's `unload()`/`load()` cycle. Chromium has no reason to exit fullscreen because the fullscreen element (the `<div>`) is untouched. The `<video>` inside it can have its source freely cleared and reassigned.

One important detail: when fullscreen is on a `<div>` rather than the `<video>`, you **lose the browser's native video fullscreen controls** (the special overlay UI Chromium provides for fullscreen `<video>`). In your case this is likely fine or even desirable since you're building a custom player UI in Electron anyway. If you need custom controls, position them absolutely inside the container `<div>`.

Also, make sure Shaka Player is not itself removing and re-creating the `<video>` DOM node during `unload()`/`load()`. Normally it doesn't — it reuses the same element — but verify by checking that you always pass the same `<video>` element reference and don't call `player.detach()` or destroy/reconstruct the player instance between items.

### Regarding Your Other Questions

**Preload API** — No, this won't help with fullscreen retention. Preloading is about buffering the next item's segments in advance to reduce transition latency. It's orthogonal to the fullscreen issue. Useful for gapless playback, but it doesn't change the DOM/source lifecycle that triggers the fullscreen exit.

**`SwitchingToNewStream` event** — This is an informational event. You could theoretically listen for it and re-request fullscreen, but `requestFullscreen()` requires a user gesture in most browsers, so you can't reliably call it from an async event handler. In Electron you *could* potentially bypass this restriction via `webPreferences`, but it would be a fragile hack compared to the wrapper `<div>` approach.

**Queue Manager** — The queue manager PR is about managing playlist state (ordering, advancement, repeat modes). It doesn't change the underlying `unload()`/`load()` mechanism for media transitions, so it wouldn't solve the fullscreen problem by itself. It would still need to be combined with the container-level fullscreen approach.

### Summary

The wrapper `<div>` approach is the clean, reliable, spec-compliant solution. The other avenues you mentioned either don't address the core issue or would require fragile workarounds.

**Comment by loicraux** [comment]
I see there is the `fullScreenElement` configuration option, that allows to configure the DOM element on which fullscreen will be done. But it already defaults to the Shaka Player Container... So that’s probably not the right approach...

Now I see this code in [shaka.ui.Controls constructor, line 402](https://shaka-player-demo.appspot.com/docs/api/ui_controls.js.html#line402) :

```javascript
this.eventManager_.listen(this.player_, 'unloading', () => {
  if (this.ad_) {
    return;
  }
  this.adCuePoints_ = [];
  this.lastSelectedTextTrack_ = null;
  if (this.isFullScreenEnabled()) {
    this.exitFullScreen_();  // <============
  }
  if (this.isPiPEnabled()) {
    this.togglePiP();
  }
  if (this.chapters_.length) {
    this.chapters_ = [];
    this.dispatchEvent(new shaka.util.FakeEvent('chaptersupdated'));
  }
});
```

Could this be the cause ? (related to https://github.com/shaka-project/shaka-player/pull/8848 )

**EDIT:** I confirm that without PR 8848, it works fine as I expect (remains in Fullscreen mode when transitioning to the next playlist item....)

**Comment by matvp91** [comment]
Yes, https://github.com/shaka-project/shaka-player/pull/8848 is the root cause.

Your request seems reasonable and I don't think we should force the controls out of fullscreen. @avelad is on holidays, I suggest we wait for his motivation to move that behavior out of the demo app and in the controls.

**Comment by avelad** [comment]
The same issue happens with `MPD Chaining`

**Comment by loicraux** [comment]
Thanks @avelad ...

Currently, my code that plays playlists does call `unload()` on the current item being played prior to calling `load()` for the next item to be played...

Will this work with your fix ? Or do I have to change something on my side ?
